### ISSUE TITLE
Staged entities do not get cleaned up

### ISSUE THREAD

**Issue by SanderMertens** [SanderMertens/flecs#53]
By @teksoc-1:

I was using the profiler tool in Xcode, to see memory behavior when creating N entities and deleting the same N entities in while ecs_progress() loop, I observed memory consumption continue to grow and what appeared to be a leak? A little further digging, it seemed that perhaps the staged columns for the data stage where not being freed; I applied a small patch of code in merge_commits() under stage.c to iterate over the staged columns of the data stage and apply ecs_vector_free() do the .data of each staged column for the merged entities. While that seemed to fix the perceived memory leak ( at least to the extent that it gets a nice happy green check mark under Xcode instruments, and I don't visually see the memory continue to grow ), I have no idea if that is even correct, or any unintended side effects I'm not considering, or if there even was a memory leak in the first place ( again, something wrong with my setup? ).

**Comment by SanderMertens** [comment]
I may not be cleaning up these arrays, as it can slow applications down if entities are continuously are being created & deleted. It shouldn't leak though. I'm currently working on redesigning parts of this code, so I'll test for this once that's done.

**Comment by teksoc-1** [comment]
briefly looking at the code, that's what it looked like, reuse of the arrays, so was expecting a spike, then for things to hold steady, but it kept growing.

**Comment by teksoc-1** [comment]
(should have perhaps followed up the dialog here, but touched on this subject on our dialog under #52 )

**Comment by teksoc-1** [comment]
of relevance, I am also seeing leaks that appear to not get cleaned up post ecs_fini(), but that is relatively less concern ( for me at least, and probably most I presume ) ... I can however provide more info on this if you want to capture in a backlog, etc.

**Comment by SanderMertens** [comment]
Some memory doesn't get cleaned up yet in `ecs_fini`, but it should not "leak" (that is, steadily increase over time). I've been holding off on making flecs valgrind clean until the data structures have stabilized- which should be in a couple of weeks.

Leaks should be fixed though, as they make the framework unusable.

**Comment by SanderMertens** [comment]
@teksoc-1 I was able to reproduce the leak and checked in a fix. I also cleaned up a few other data structures that leaked on `ecs_fini` (but did not cause leaks while running). Not all of them are gone, but valgrind looks much cleaner now.

The one leak that remains (not steadily increasing, but in `ecs_fini`) is complex to free, and is in a part of the code that is completely redesigned in the type_graph branch. I'll probably won't fix that one, as it will get fixed automatically when the branch is merged.
